### DISCUSSION TITLE
RVE with multiple small grains embedded in a big grain

### DISCUSSION THREAD

**Discussion by danishTUE** [neperfepx/neper#549] [Q&A] [answered]
Hello Romain,

I am interested to create a RVE/Unit-cell where multiple small grains are embedded in a region of radius r in a bigger grain. How can I create such an RVE with NEPER or trick Neper to create and mesh such an  RVE? I am attaching a schematic figure with this post (although the smaller grain shown in the schematic are spherical but they can be normal tessellation as well)

Regards,

Danish
![RVE](https://user-images.githubusercontent.com/95438134/187080195-252bd439-c7a4-4f94-81f8-af5bc203e5a7.PNG)

**Comment by danishTUE** [comment]
Ofcourse, one way would be to create a cube of grains and treat the grains outside the region of 2r with same orientation. But since the size ratio of bigger grain to smaller grains is pretty large, its taking a lot of time create domain with so many small size grains. I am trying to figure out if there exists a faster and smarter way. Thanks!

**Reply by mkasemer** [reply]
In a similar vein, could you create a multilevel tessellation where:
- the first level has a relatively small number of grains (perhaps 30)
- the second level sub-discretizes all grains with 1 subdomain, except for the center grain which you can discretize with many subgrains
- all of the first level grains that were sub-discretized using 1 grain (i.e., no subgrains) are assigned the same orientation.

Would this do the trick?

**Reply by danishTUE** [reply]
Thanks Matthew. Yes I think this should serve the purpose. I saw this is in some sense discussed here as well: https://github.com/neperfepx/neper/discussions/401
However, for this method, going by the example given by Romain in this post, its not very clear how we can control which grain to be sub-discretized (for .e.g in my case it would be the grain whose centroid is closest to centroid of the domain). One way would be to run the tessellation run two times. Once without subdiscretization just to find the grain# of interest. Then run again with subdiscretization with grain of interest pointed via the msfile. Can there be a control on the number of subdiscretization of a grain, let say based on their centroid position, without having to run the tesselation two times?

**Reply by rquey** [reply]
Probably not, but I think that the most significant difference between Matt's solution and mine (below) is the region formed by the embedded grains - either a a regular and convex polyhedron or an irregular sphere. Depends on what you want.

My solution also needs 2 steps...

**Comment by rquey** [comment, marked as answer]
What is your `-morpho`?

What I would probably do is generate the tessellation with grain 1 undersized, and then recompute the tessellation in a larger domain:

```
neper -T -n 100 -domain "circle(0.3):translate(-0.15,-0.15)" -dim 2 -o tess1
neper -T -n 100 -domain "square(1,1):translate(-0.5,-0.5)" -dim 2 -morphooptiini "file(tess1.tess)" -o tess2
neper -V tess1.tess -dataedgerad 0.002 -cameracoo x:y:2.5 -imagesize 800:800 -print img1
neper -V tess2.tess -dataedgerad 0.002 -datacellcol "int:((x*x+y*y>0.1*0.1)?0:id)" -cameracoo x:y:2.5 -imagesize 800:800 -print img2
```
![image](https://user-images.githubusercontent.com/2659376/187169734-c522dcf7-3a42-4a1c-b283-58dc62fb1a03.png)
![image](https://user-images.githubusercontent.com/2659376/187169750-2b85c6d0-92b8-4d5e-8f91-8b5369714170.png)

**Reply by danishTUE** [reply]
Ideally the morpho for smaller grains would be diameq with some lognormal distribution. But I can also live with equal grain sizes. Based on the figures it looks what I want. I am not fully understanding what is exactly happening in your example. How is recomputing the tesselation in a larger square domain leading to this type of RVE? Will it also work for 3D? Thanks!

**Reply by danishTUE** [reply]
If I am understanding correctly, in this method there is no control on the size/volume of the region which would maintain their original small size (I see that the outer grains of the original circle are stretched in square domain)?

**Reply by rquey** [reply]
The first command is computing the tessellation into small grains in a region a little bit larger than 2r.
The second command is recomputing the tessellation in the total domain, which includes grain 1 and the set of small grains. Since the domain is larger but does not include other grains, the grains which are on the shell of the first tessellation extend "radially" to fill the domain. The small grains are colored by their ids and all cells constituting grain 1 are colored in gray. The tessellation is not visually appealing, with multiple cells constituting grain 1, but since all have the same orientation, it is not so important.

Yes, it also work in 3D.
